### PR TITLE
rbac: setup graphql schema and query for fetching roles and permissions

### LINKED ISSUE #45440: RBAC: Implement GraphQL query for fetching all available permissions

### LINKED ISSUE #45441: RBAC: Implement GraphQL query for fetching all available roles

### LINKED ISSUE #45442: RBAC: Implement GraphQL query for fetching a user's permissions

### LINKED ISSUE #45443: RBAC: Implement GraphQL query for fetching user roles

### PR DESCRIPTION
**Pull request by BolajiOlajide**
Closes #45441
Closes #45443
Closes #45442
Closes #45440

## Test plan

This turned out to be quite a large PR; I didn't anticipate it'll grow this large. This PR adds the initial graphql layer for RBAC, it provides the schema for querying the API to fetch roles, permissions (belonging to the user or in the system).

Added unit tests to ensure the additions are well tested.

### CODE REVIEW

File: internal/database/permissions.go [hunk 1]
```diff
@@ -41,8 +41,10 @@ type PermissionStore interface {
 	BulkDelete(ctx context.Context, opts []DeletePermissionOpts) error
 	// GetByID returns the permission matching the given ID, or PermissionNotFoundErr if no such record exists.
 	GetByID(ctx context.Context, opts GetPermissionOpts) (*types.Permission, error)
-	// List returns all the permissions in the database.
-	List(ctx context.Context) ([]*types.Permission, error)
+	// List returns all the permissions in the database that matches the options.
+	List(ctx context.Context, opts PermissionListOpts) ([]*types.Permission, error)
+	// Count returns the number of permissions in the database mtching the options provided.
```
Comment (courier-new) [inline internal/database/permissions.go:46]:
```suggestion
	// Count returns the number of permissions in the database matching the options provided.
```

File: internal/database/permissions.go [hunk 2]
```diff
@@ -245,25 +253,78 @@ func (p *permissionStore) GetByID(ctx context.Context, opts GetPermissionOpts) (
...
+	q := sqlf.Sprintf(
+		permissionListQueryFmtStr,
+		selects,
+		joins,
+		preds,
+	)
+
+	if opts.UserID != 0 {
+		q = sqlf.Sprintf("%s\n%s", q, sqlf.Sprintf("GROUP BY permissions.id"))
+	}
```
Comment (courier-new) [inline internal/database/permissions.go:303]: Nice, I would've forgotten about this. 😄

**Review by courier-new — APPROVED** [review]
Looks great overall!

One thing, I see you've chosen to implement the resolver so that it's only available on enterprise instances. I'm wondering if we'd actually prefer it to just always be available, similarly to how we decided to seed the default roles for every instance? Even if we're primarily only concerned with using them for enterprise features like Batch Changes, we also talked about wanting to query permissions from the root `SourcegraphWebApp` in the frontend web client, and I feel like that would be more straighforward/easier to reason about if the resolver is always available. Are there costs or concerns related to it being available in OSS/dotcom or the like that you've thought about?

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 3]
```diff
@@ -380,31 +380,35 @@ func prometheusGraphQLRequestName(requestName string) string {
...
+	return NewSchema(db, gitserver.NewClient(db), nil, nil, nil, nil, nil, license, nil, nil, nil, nil, nil, nil, nil)
 }
 
 func NewSchemaWithWebhooksResolver(db database.DB, webhooksResolver WebhooksResolver) (*graphql.Schema, error) {
-	return NewSchema(db, gitserver.NewClient(db), nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, webhooksResolver)
+	return NewSchema(db, gitserver.NewClient(db), nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, webhooksResolver, nil)
+}
+
+func MewSchemaWithRBACResolver(db database.DB, rbacResolver RBACResolver) (*graphql.Schema, error) {
+	return NewSchema(db, gitserver.NewClient(db), nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, rbacResolver)
```
Comment (kopancek) [inline cmd/frontend/graphqlbackend/graphqlbackend.go]: At this point, it might be worth refactoring this interface. Providing a long list of `nil` arguments is not really readable to be honest.

But since this PR is already huge, maybe we can do it in a followup.

File: cmd/frontend/graphqlbackend/rbac.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+type PermissionConnectionResolver interface {
+	Nodes(ctx context.Context) ([]PermissionResolver, error)
+	TotalCount(ctx context.Context) (int32, error)
+	PageInfo(ctx context.Context) (*graphqlutil.PageInfo, error)
+}
+
+type RBACResolver interface {
+	// MUTATIONS
+
+	// QUERIES
```
Comment (kopancek) [inline cmd/frontend/graphqlbackend/rbac.go:43]: Since we only seem to have queries for now, do we need these comments?

File: cmd/frontend/graphqlbackend/rbac.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	NodeResolvers() map[string]NodeByIDFunc
+}
+
+type ListRoleArgs struct {
+	User  *graphql.ID
+	First int32
+	After *string
+}
+
+func (l ListRoleArgs) LimitOffset() (*database.LimitOffset, error) {
```
Comment (kopancek) [inline cmd/frontend/graphqlbackend/rbac.go:57]: Instead of limit/offset, should we use the new pagination mechanism that @thenamankumar was working on?

File: cmd/frontend/graphqlbackend/rbac.graphql [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    """
+    id: ID!
+    """
+    The human readable name for this role.
+    """
+    name: String!
+    """
+    This indicates whether a role can be modified or deleted.
+    """
+    system: Boolean!
```
Comment (kopancek) [inline cmd/frontend/graphqlbackend/rbac.graphql:16]: Should we call this parameter `readOnly` since that is actually what it means? I can imagine having other possibilities of readonly roles in the future, not just `system` roles.

File: cmd/frontend/graphqlbackend/rbac.graphql [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        Only return N records.
+        """
+        first: Int = 50
+        """
+        Opaque cursor for pagination.
+        """
+        after: String
+    ): RoleConnection!
+    """
+    The list of all permissions belonging to roles assigned to this user.
```
[the remark this answers is not in the document]
Comment (kopancek) [reply cmd/frontend/graphqlbackend/rbac.graphql]:
```suggestion
    The list of permissions granted to this user based on all of their roles.
```

File: cmd/frontend/graphqlbackend/user.go [hunk 8]
```diff
@@ -478,6 +478,18 @@ func (r *UserResolver) BatchChangesCodeHosts(ctx context.Context, args *ListBatc
...
+	id := r.ID()
+	args.User = &id
+	return EnterpriseResolvers.rbacResolver.Roles(ctx, args)
+}
+
+func (r *UserResolver) Permissions(ctx context.Context, args *ListPermissionArgs) (PermissionConnectionResolver, error) {
+	id := r.ID()
+	args.User = &id
+	return EnterpriseResolvers.rbacResolver.Permissions(ctx, args)
+}
```
[the remark this answers is not in the document]
Comment (kopancek) [reply cmd/frontend/graphqlbackend/user.go:491]: +1

File: cmd/frontend/graphqlbackend/node.go [hunk 9]
```diff
@@ -94,6 +94,16 @@ func (r *NodeResolver) ToBatchChange() (BatchChangeResolver, bool) {
...
+func (r *NodeResolver) ToRole() (RoleResolver, bool) {
+	n, ok := r.Node.(RoleResolver)
+	return n, ok
+}
+
+func (r *NodeResolver) ToPermission() (PermissionResolver, bool) {
+	n, ok := r.Node.(PermissionResolver)
+	return n, ok
+}
+
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/node.go:106]: Sure thing.

File: enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+	"strconv"
+	"sync"
+
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend/graphqlutil"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+)
+
+var _ graphqlbackend.PermissionConnectionResolver = &permissionConnectionResolver{}
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go:14]: Why is this needed?

File: enterprise/cmd/frontend/internal/rbac/resolvers/permissions.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+		return nil, ErrIDIsZero{}
+	}
+
+	permission, err := r.db.Permissions().GetByID(ctx, database.GetPermissionOpts{
+		ID: permissionID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &permissionResolver{permission: permission}, nil
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/rbac/resolvers/permissions.go:28]: The pattern I have seen elsewhere was to call a function `NewPermissionResolver(permission)` here and it would give us `&permissionResolver{permission: permission}` as it's output.

File: enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+	// cache results because they are used by multiple fields
+	once        sync.Once
+	permissions []*types.Permission
+	totalCount  int
+	err         error
+}
+
+func (r *permissionConnectionResolver) TotalCount(ctx context.Context) (int32, error) {
+	_, count, err := r.compute(ctx)
```
Comment (kopancek) [inline enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go:28]: All of these resolver methods should only be accessible by site-admin and the user in question. Can you add guards for that?

This is true for all the resolvers that are part of RBAC and all of their mutations/queries.

File: internal/database/permissions.go [hunk 13]
```diff
@@ -41,8 +41,10 @@ type PermissionStore interface {
 	BulkDelete(ctx context.Context, opts []DeletePermissionOpts) error
 	// GetByID returns the permission matching the given ID, or PermissionNotFoundErr if no such record exists.
 	GetByID(ctx context.Context, opts GetPermissionOpts) (*types.Permission, error)
-	// List returns all the permissions in the database.
-	List(ctx context.Context) ([]*types.Permission, error)
+	// List returns all the permissions in the database that matches the options.
```
Comment (eseliger) [inline internal/database/permissions.go:44]: some tests for the new methods/options in here would be great!

File: enterprise/cmd/frontend/internal/rbac/resolvers/role_connection_test.go [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+		var response struct{ Node apitest.User }
+		apitest.MustExec(actorCtx, t, s, input, &response, listUserRoles)
+
+		if diff := cmp.Diff(want, response.Node); diff != "" {
+			t.Fatalf("wrong role response (-want +got):\n%s", diff)
+		}
+	})
+
+	t.Run("listing a user's roles (site admin)", func(t *testing.T) {
+		t.Skip()
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/rbac/resolvers/role_connection_test.go:166]:
```suggestion
```

?

File: enterprise/cmd/frontend/internal/rbac/resolvers/permissions.go [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+		opts.RoleID = roleID
+	}
+
+	if args.User != nil {
+		userID, err := gql.UnmarshalUserID(*args.User)
+		if err != nil {
+			return nil, err
+		}
+
+		opts.UserID = userID
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/rbac/resolvers/permissions.go:50]: null ID check

File: enterprise/cmd/frontend/internal/rbac/resolvers/permission.go [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+func marshalPermissionID(id int32) graphql.ID { return relay.MarshalID(permissionIDKind, id) }
+
+func unmarshalPermissionID(id graphql.ID) (permissionID int32, err error) {
+	err = relay.UnmarshalSpec(id, &permissionID)
+	return
+}
+
+func (r *permissionResolver) ID() graphql.ID {
+	// 🚨 SECURITY: This needs to be the marshalled.
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/rbac/resolvers/permission.go:28]: how is this a security thing?

File: cmd/frontend/graphqlbackend/rbac.graphql [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    """
+    name: String!
+    """
+    Indicates whether a role is a default system role, which cannot be modified or deleted, or a custom role added by a site admin.
+    """
+    system: Boolean!
+    """
+    The list of permissions that will be granted to any user with this role.
+    """
+    permissions: PermissionConnection!
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/rbac.graphql:20]: should this have `first`, `after` arguments?

File: cmd/frontend/graphqlbackend/rbac.graphql [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+"""
+type Permission implements Node {
+    """
+    The globally unique identifier for this permission.
+    """
+    id: ID!
+    """
+    The namespace in which this permission belongs to.
+    """
+    namespace: String!
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/rbac.graphql:74]: should this be an enum?

File: cmd/frontend/graphqlbackend/rbac.graphql [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+        """
+        first: Int = 50
+        """
+        Opaque cursor for pagination.
+        """
+        after: String
+        """
+        Only return permissions associated with the role with this ID.
+        """
+        role: ID
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/rbac.graphql:119]: should this be 

```graphql
query {
 node(id: roleID) {
 ... on Role {
    permissions(first: $first, after: $after) { nodes { name } } 
  }
}
}
```

vs an ID property on the global list? Feels a bit more graphql-y to me

Comment (eseliger) [reply cmd/frontend/graphqlbackend/rbac.graphql:119]: probably the same applies to `user`

**Review by eseliger — APPROVED** [review]
good work, this feels very clean to me!

File: cmd/frontend/graphqlbackend/user.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/user.go:491]: These tests already exist - I added them in the graphql resolver tests for [roles](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=bo%2Fadd-graphql-type-role-permissions&file=enterprise%2Fcmd%2Ffrontend%2Finternal%2Frbac%2Fresolvers%2Frole_connection_test.go&editor=VSCode&version=2.2.14&start_row=111&start_col=0&end_row=204&end_col=1) and [permissions](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@bo/add-graphql-type-role-permissions/-/blob/enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection_test.go?L112%3A1-216%3A2).

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 3 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Yeah, discussed something we can do about with Erik earlier today.

File: cmd/frontend/graphqlbackend/rbac.go (hunk 4 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/rbac.go:43]: We'll have mutations in the next PR. This is just a precursor to the next PR.
I had to remove some of the other changes to make this PR smaller.

File: cmd/frontend/graphqlbackend/rbac.graphql (hunk 6 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/rbac.graphql:16]: We decided to go with `System` to represent the fact that these roles come with any Sourcegraph instance and are used internally; at some point, we might have a concept of read-only roles but those won't necessarily be the same as the current system roles `USER` and `SITE_ADMINISTRATOR` so using the name `system` gives a clear distinction.

File: enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go (hunk 10 above)
Comment (BolajiOlajide) [reply enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go:14]: To ensure that `permissionConnectionResolver` struct implements the `graphqlbackend.PermissionConnectionResolver` interface.

File: enterprise/cmd/frontend/internal/rbac/resolvers/permissions.go (hunk 11 above)
Comment (BolajiOlajide) [reply enterprise/cmd/frontend/internal/rbac/resolvers/permissions.go:28]: Yeah, it's the same. One is just easier to mock out and use in tests.

File: cmd/frontend/graphqlbackend/user.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (courier-new) [reply cmd/frontend/graphqlbackend/user.go:491]: Ah okay perfect, missed that. Thanks Bolaji!

File: cmd/frontend/graphqlbackend/rbac.graphql (hunk 17 above)
Comment (courier-new) [reply cmd/frontend/graphqlbackend/rbac.graphql:20]: I think Bolaji may have intentionally omitted them here to just always fetch _all_ permissions, given the list of permissions for a given role will be relatively small (especially right now when there's only a handful of permission types). From the previews of the site admin UI I've seen, it would be very cumbersome and clunky to need to paginate them.

File: cmd/frontend/graphqlbackend/rbac.go (hunk 5 above)
Comment (thenamankumar) [reply cmd/frontend/graphqlbackend/rbac.go:57]: @kopancek is right. Lets have cursor-based pagination here. You can read the docs here: https://docs.sourcegraph.com/dev/background-information/web/graphql#best-practices-for-paginating-resources-with-graphql

Let me know if you would like to pair on this.

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 3 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: Yeah, I'll try and get to that in my spare time. I discussed with @eseliger an excellent approach to tackle this.

File: cmd/frontend/graphqlbackend/rbac.go (hunk 4 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/rbac.go:43]: This PR is a pre-cursor to another PR that adds the mutation. Left it there because it makes the code readable.

File: enterprise/cmd/frontend/internal/rbac/resolvers/roles.go [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+	var err error
+	var opts = database.RolesListOptions{}
+
+	if args.User != nil {
+		userID, err := gql.UnmarshalUserID(*args.User)
+		if err != nil {
+			return nil, err
+		}
+
+		opts.UserID = userID
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply enterprise/cmd/frontend/internal/rbac/resolvers/roles.go:24]: Nice catch. I'll add that.

File: enterprise/cmd/frontend/internal/rbac/resolvers/permission.go (hunk 16 above)
Comment (BolajiOlajide) [reply enterprise/cmd/frontend/internal/rbac/resolvers/permission.go:28]: Oh, it's not. It is an oversight.

File: cmd/frontend/graphqlbackend/rbac.graphql (hunk 17 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/rbac.graphql:20]: Yes, that's correct. I'll add a comment to make this more explicit in the codebase.

File: cmd/frontend/graphqlbackend/rbac.graphql (hunk 19 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/rbac.graphql:119]: Yeah, you're right. That's more graphql-y.

File: enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go (hunk 12 above)
Comment (BolajiOlajide) [reply enterprise/cmd/frontend/internal/rbac/resolvers/permission_connection.go:28]: Added guards in place for this. Only a user / site admin can fetch a user's roles or permissions.

File: internal/database/permissions.go (hunk 13 above)
Comment (BolajiOlajide) [reply internal/database/permissions.go:48]: Added.

File: internal/database/roles.go [hunk 21]
```diff
@@ -61,6 +61,7 @@ type (
 type RolesListOptions struct {
 	*LimitOffset
 	System bool
+	UserID int32
```
[the remark this answers is not in the document]
Comment (BolajiOlajide) [reply internal/database/roles.go:65]: Added.

File: cmd/frontend/graphqlbackend/rbac.graphql (hunk 18 above)
Comment (BolajiOlajide) [reply cmd/frontend/graphqlbackend/rbac.graphql:91]: It can. Not sure if there's a lot of benefit to it being an `enum` yet as it'd mean every time a namespace is added, we'll have to update the `enum`, so I'll monitor and see if there's a benefit to this.
